rfc-156-auto-merge-internal-prs.md

```diff
@@ -0,0 +1,210 @@
+
+Note that the '95% code coverage' figure is a guide, which is not programmatically enforced. It is also not necessarily actively monitored after continuous deployment is enabled for an app, so could degrade over time. Approving this RFC means accepting that it is ever more important to maintain high automated testing standards.
+
+[RFC-128]: https://github.com/alphagov/govuk-rfcs/blob/main/rfc-128-continuous-deployment.md
+
+#### 3. CI build passes
+
+A dependency must not be automatically merged if an associated CI build fails.
+
+This follows on from the criterion above: continuously deployed apps have good test coverage, so it makes sense to check the outcome of the CI build, only merging the PR if all tests pass.
+
+#### 4. Version increase is patch or minor
+
+A 'major' version increase must not be automatically merged, even if the PR checks pass, as it could require some additional manual configuration before merging. This risk is lower with patch or minor versions.
+
+One could argue that, so long as the tests pass, even major version upgrades should be automatically merged. But these often have nuances and recommendations that are not necessarily covered by unit tests. To illustrate the point - even though it's not an internal dependency - a classic example is Rails, where [manual steps are required][rails-major-upgrade], even if the CI build passes. It is not worth risking automatic merging, versus the time it would take for a developer to manually process the changes. It could be argued that developers should be fully aware whenever their app is moved to a major new dependency version.
+
+Note that this does put extra onus on getting our [semantic versioning][semver] right. Before, it was just semantics, but if this RFC is approved, a developer's choice of versioning will determine whether or not a dependency will be automatically rolled out. That said, we do have existing [guidance on versioning our ruby gems][govuk-semver-guidance].
+
+[govuk-semver-guidance]: https://docs.publishing.service.gov.uk/manual/publishing-a-ruby-gem.html#versioning
+[rails-major-upgrade]: https://docs.publishing.service.gov.uk/manual/keeping-software-current.html#rails
+[semver]: https://semver.org/
+
+#### 5. During office hours
+
+We want to limit the times at which auto-merging can take place, to typical office hours only. If a gem with a bug were to be released at 5pm on a Friday, this wouldn't be deployed into the apps until after most people have stopped working for the week, potentially creating work for those on-call if there is a user-facing impact.
+
+We propose a broad auto-merge window of between 9am and 4pm, Monday to Friday. Auto-merging must not happen outside of those times. In practice, the auto-merging will actually happen in a much narrower time period, around 9:30am, for reasons outlined below.
```

Comment:
I wonder whether an earlier time on Friday might be sensible... or just leaving Friday out entirely because of its proximity to the weekend...

Reply:
+1 to leaving out Fridays.

Reply:
@richardTowers the RFC currently proposes that Dependabot would run around 9:30am every weekday, i.e. auto merging should happen in the morning (assuming Dependabot runs at the specified time and that the GitHub Actions runs aren't delayed). I appreciate the wording of 9am-4pm doesn't help, here (I didn't see the need to constrain developers to only running Dependabot at 9:30am when we really only want to ensure it happens within office hours, but this does give unnecessary freedom and there's no obvious reason why they might want to schedule Dependabot to run later).

I've clarified the language so that it's explicitly proposing Dependabot running 9:30am Monday-Friday.

Both of you, let me know if you still think Fridays should be left out. Happy to apply that change to the RFC, but my two cents are that it's not particularly riskier than any other weekday and a bit arbitrary to drop (e.g. why not drop Thursday too?). We should have enough confidence in the other safeguards and the several hours of rollback time (if needed), IMO.